### ISSUE TITLE
sql: plan gist decoder panics when table in plan has been dropped

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/cockroach#83537]
Here's a logic test showing the panic:

```
statement ok
CREATE TABLE t (k INT PRIMARY KEY, a INT)

let $gist
EXPLAIN (GIST) INSERT INTO t VALUES (1, 10)

statement ok
DROP TABLE t

statement ok
SELECT crdb_internal.decode_plan_gist('$gist')
```

The stack trace looks like:

```
* ERROR: a panic has occurred!
* runtime error: invalid memory address or nil pointer dereference
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:755
*   | [...repeated from below...]
* Wraps: (2) while executing: SELECT crdb_internal.decode_plan_gist(_)
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:755
*   | runtime.gopanic
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:1038
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58
*   | runtime.gopanic
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:1038
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58
*   | runtime.gopanic
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:1038
*   | runtime.panicmem
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:221
*   | runtime.sigpanic
*   |   /usr/local/opt/go/libexec/src/runtime/signal_unix.go:735
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*emitter).emitNodeAttributes
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:755
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:53
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:66
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.DecodePlanGistToRows
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/plan_gist_factory.go:153
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).DecodeGist                                                                                             *   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:694
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.(*gistPlanGenerator).Start
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/generator_builtins.go:505
*   | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).nextInputRow
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:171
*   | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*projectSetProcessor).Next
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/project_set.go:254
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:231
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*invariantsChecker).Next
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:93
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:99
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:107
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:242
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).next
*   |   /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:246
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run
```

**Comment by joshimhoff** [comment]
This panic also didn't get caught via a deferred call to `logcrash` (https://github.com/cockroachdb/cockroach/tree/master/pkg/util/log/logcrash), so it and/or similar issues won't appear in Sentry and/or non-stderr logs (which in CC we deliver to Splunk) . Can we fix that too, if it's ~easy?

**Comment by mgartner** [comment]
Here's some mitigations for those that run into this issue, until v22.1.4 is released with the fix:

1. Clear all statement stats with `SELECT crdb_internal.reset_sql_stats()`. This will delete historical statement stats data (everything on the statement statistics page in the console UI). This is only a temporary mitigation because if indexes or tables are dropped in the future, the panics can return.
2. Disable plan gists with the session setting `SET disable_plan_gists=true`. To disable plan gists by default for all newly created sessions, use `ALTER ROLE ALL SET disable_plan_gists = true`. This will disable new query plans from being shown in the statement details page in the console UI. Due to another bug, #82095, this setting should not be used when running v22.1.0 or v22.1.1.

The more surefire way to prevent these panics when running v22.1.2 or v22.1.3 is to first disable new plan gists from being created and then clear all statement statistics. Once the upgrade to v22.1.4 is complete, plan gists can be re-enabled. For example:

```sql
-- 1. Disable plan gists.
ALTER ROLE ALL SET disable_plan_gists = true;

-- 2. Cycle all connections to ensure the setting propagates to all sessions.

-- 3. Clear statement statistics.
SELECT crdb_internal.reset_sql_stats();

-- 4. Later, upgrade to v22.1.4, then re-enable plan gists.
ALTER ROLE ALL SET disable_plan_gists = true;

-- 5. Cycle all connections to ensure the setting propagates to all sessions.
```
